### PR TITLE
Port admin IPs (RIPD-820)

### PR DESCRIPTION
**Pull request by miguelportilla**
* Config file rpc_admin_allow section is deprecated.
* Port setting 'admin' only accepts allowable admin IP addresses.
* Admin IP addresses can be specified as comma separated values.
* 127.0.0.1 is no longer a default admin IP.
* 0.0.0.0 may be specified. Cannot be combined with other IP addresses.

Reviewers: @nbougalis @seelabs

### CODE REVIEW

File: src/ripple/server/Port.h [hunk 1]
```diff
@@ -97,11 +98,16 @@ inline
...
-        (p.allow_admin ? ", admin, " : ", ") <<
-        p.protocols() << ")";
+    os << "'" << p.name << "' (ip=" << p.ip << ":" << p.port << ", ";
+
+    if (! p.admin_ip.empty ())
+    {
+        os << "admin IPs:";
+        for (auto const& ip : p.admin_ip)
+            os << ip.to_string () << ", ";
+    }
```
Comment (seelabs) [inline src/ripple/server/Port.h]: It's a shame C++ doesn't have a nice way to stream a comma separated list. I don't object to the extra comma (unless there's a terse way that you can think of).

File: src/ripple/websocket/Connection.h [hunk 2]
```diff
@@ -262,8 +262,7 @@ Json::Value ConnectionImpl <WebSocket>::invokeCommand (Json::Value& jvRequest)
     Json::Value jvResult (Json::objectValue);
 
     auto required = RPC::roleRequired (jvRequest[jss::command].asString());
-    Role const role = requestRole (required, port_, jvRequest, m_remoteAddress,
-                                   getConfig().RPC_ADMIN_ALLOW);
+    Role const role = requestRole (required, port_, jvRequest, m_remoteAddress);
```
Comment (seelabs) [inline src/ripple/websocket/Connection.h]: Feel like renaming `port_` to `m_port` while you're in there? (I know it's not new. It's private and only used a couple place and it's good hygiene.)

**Comment by seelabs** [comment]
I did a quick pass and this looks good. I'll do a more careful pass later, but I suspect there won't be much to change in this.

**Comment by nbougalis** [comment]
This looks good to me: :+1:. 

I'm not super-hot on the `admin = [ IP, IP, IP, ... ]` syntax since it's unique to this block of code, but I guess such are limitations of ini-style files.

File: src/ripple/server/impl/Role.cpp [hunk 3]
```diff
@@ -38,29 +38,28 @@ passwordUnrequiredOrSentCorrect (HTTP::Port const& port,
...
 bool
-ipAllowed (beast::IP::Endpoint const& remoteIp,
-           std::vector<beast::IP::Endpoint> const& adminAllow)
+ipAllowed (beast::IP::Address const& remoteIp,
+           std::vector<beast::IP::Address> const& adminIp)
 {
-    return std::find(adminAllow.begin(), adminAllow.end(),
-                     remoteIp.at_port(0)) != adminAllow.end();
+    return ! adminIp.empty () && std::find_if (adminIp.begin (), adminIp.end (),
+        [&remoteIp](beast::IP::Address const& ip) { return ip.is_any () ||
```
Comment (seelabs) [inline src/ripple/server/impl/Role.cpp]: Why is the `! adminIp.empty ()` needed? It looks like it works the same without it.

File: src/ripple/core/ConfigSections.h [hunk 4]
```diff
@@ -55,7 +55,6 @@ struct ConfigSection
 #define SECTION_PATH_SEARCH_MAX         "path_search_max"
 #define SECTION_PEER_PRIVATE            "peer_private"
 #define SECTION_PEERS_MAX               "peers_max"
-#define SECTION_RPC_ADMIN_ALLOW         "rpc_admin_allow"
```
Comment (seelabs) [inline src/ripple/core/ConfigSections.h]: Should we warn if the config file contains an obsolete `rpc_admin_allow` section? I'm thinking that would be a different PR (warn if the config file contains unknown sections).

File: src/ripple/server/Port.h (hunk 1 above)
Comment (miguelportilla) [inline src/ripple/server/Port.h]: There will not be be an extra comma since the 'protocol' is added afterward.

File: src/ripple/server/impl/ServerHandlerImp.cpp [hunk 5]
```diff
@@ -521,7 +520,7 @@ struct ParsedPort
 
     boost::optional<boost::asio::ip::address> ip;
     boost::optional<std::uint16_t> port;
-    boost::optional<bool> allow_admin;
+    boost::optional<std::vector<beast::IP::Address>> admin_ip;
```
Comment (seelabs) [inline src/ripple/server/impl/ServerHandlerImp.cpp]: Does it make sense to remove the `boost::optional` and just use `std::vector`? I don't see in our code where a `false` boost optional here would act differently from an empty vector (tho I could have missed it).

File: src/ripple/websocket/Connection.h (hunk 2 above)
Comment (miguelportilla) [inline src/ripple/websocket/Connection.h]: Good catch, I'll change this to match the class style. Someone must have added port_ at a later date.

File: src/ripple/core/ConfigSections.h (hunk 4 above)
Comment (miguelportilla) [inline src/ripple/core/ConfigSections.h]: We've not warned in the past when deprecating config sections. Its probably a good idea to warn as you suggested (separate PR). I did remove rpc_admin_allow from the example config file, added new docs on the admin setting.

File: src/ripple/server/impl/ServerHandlerImp.cpp (hunk 5 above)
Comment (miguelportilla) [inline src/ripple/server/impl/ServerHandlerImp.cpp]: This vector will not be allocated if no admin setting is specified. Since the setting is optional and the variable was an optional variable before, I kept it as such.

Comment (miguelportilla) [inline src/ripple/server/impl/ServerHandlerImp.cpp]: @nbougalis The comma separated syntax is not unique to this config element. The protocol is specified exactly the same way.

**Comment by seelabs** [comment]
The Visual Studio projects change if I run `scons vcxproj`. However, last I checked the vcxproj was stale on develop. So unless you made a change what requires a new vcxproj (and I don't see where you did) you don't need to do anything.

File: src/ripple/server/impl/Role.cpp (hunk 3 above)
Comment (miguelportilla) [inline src/ripple/server/impl/Role.cpp]: Not needed, will remove. Old optimizing habits.

**Comment by miguelportilla** [comment]
@nbougalis The comma separated syntax is not unique to this config element. The protocol is specified exactly the same way.

File: src/ripple/server/impl/ServerHandlerImp.cpp (hunk 5 above)
Comment (nbougalis) [inline src/ripple/server/impl/ServerHandlerImp.cpp]: I thought square brackets where needed. Did I misinterpret the syntax shown?

Comment (miguelportilla) [inline src/ripple/server/impl/ServerHandlerImp.cpp]: @nbougalis  The syntax I wrote in the docs was modeled after the protocol section. In both cases brackets are optional, either way does not affect protocols or IPs specified.

Comment (nbougalis) [inline src/ripple/server/impl/ServerHandlerImp.cpp]: Ah, alright. I see it now. Thanks.

**Comment by seelabs** [comment]
:+1:

**Comment by rec** [comment]
Sorry, this didn't get merged because there was a conflict!  :-)

**Comment by sublimator** [comment]
I messed up the naming of the file for the uniport tests, so we'll need the contents of these commits too: https://github.com/sublimator/rippled/commits/uniport-test-update

I updated the tests for you.

**Comment by miguelportilla** [comment]
@sublimator Thanks!

@rec rebased and cherry picked Nicholas's commits

**Comment by nbougalis** [comment]
Merged as 6d79004d4f5ed8422bfbba87251956bd595950d5
